sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/RecordReaderIterator.scala

```diff
@@ -27,7 +29,8 @@ import org.apache.spark.sql.catalyst.InternalRow
  * Note that this returns [[Object]]s instead of [[InternalRow]] because we rely on erasure to pass
  * column batches by pretending they are rows.
  */
-class RecordReaderIterator[T](rowReader: RecordReader[_, T]) extends Iterator[T] {
+class RecordReaderIterator[T](
+    private[this] var rowReader: RecordReader[_, T]) extends Iterator[T] with Closeable {
```

Comment:
By nulling out the `rowReader` I think that this will prevent memory consumption from becoming too high in the list of task completion callbacks.